### PR TITLE
Merge Feature/car 67 with master

### LINKED ISSUE #61: Navigate around obstacles
### Description
As a driver, I want the car to be able to navigate around obstacles, so that if I tell the car to go somewhere, the car will be able to fullfill my request regardless if there are obstacles on the way or not.

### Acceptance criteria
Upon meeting an obstacle, the car should navigate its way around it automatically 95% of the time.

### PR DESCRIPTION
**Pull request by majawennroth**
#67 Implement an almost always coopperative drive around obstical method.

### CODE REVIEW

**Comment by majawennroth** [comment]
[For Dimitrios.pdf](https://github.com/DIT112-V20/group-07/files/4603495/For.Dimitrios.pdf)
In this document you will find a description of our questions and the problem we are facing with the code. we would appriciate if you could have a look at it and that we can discuss it on monday during the meeting. @platisd 

Using:
* `Smartcar.h`
* Ultrasonic sensor in the front and on the right side
* Micro LIDAR on the left side.
* Odometers to calculate distance moved

What we would like you to look at is the `driveAroundObstical()`, the `obstacleTurnLeft()` and
the `obstacleTurnRight()` in the `car.ino` file. Because the car is not always doing what we
expect it to do.

Three main problems:
1. When you first turn on the car and press start it wont react to obstacles in front of it.
instead it drives straight into the obstacle and after about 10s it either stop or turn.
Once the car has stopped and you press start again it will react as expected.
2. Looking at the picture, once the car has gone into the first state it will continue
through state, 1, 2, 3, 4, 5, 6 and 7. It then jumps to state 10 and stops there. It
seems like it skips rows in the sketch. (this happened almost every time but not
always )
3. The car does not always turn 90 degrees, sometimes a bit more and sometimes a
little less which make the car drive into the obstacle again.
We want the car to drive around an obstacle (look at picture further down):

FYI:
* The straight arrows are indicating that the car is moving forward
* The turned arrows are indicating that the car is turning in that direction
* The dots are indicating stop
* The squares are illustrating the obstacle.

96259165_2719589784937652_947564863188631552_n

**Comment by platisd** [comment]
@majawennroth I added the contents of the PDF in your response. When we discuss code or the project we should be doing it in the appropriate channels, which is here. The Monday meeting is not the appropriate setting for such discussions as it is on one hand very short on the other it does not give us immediate access to the code like GitHub does.

**Comment by majawennroth** [comment]
Okay, but when is an appropriate time then? Because we would really like to have the discussion face to face and not via text becuause it is taking a bit too long to get answers and it is easy to missunderstand each other. @platisd

**Comment by platisd** [comment]
> When you first turn on the car and press start it wont react to obstacles in front of it.
instead it drives straight into the obstacle and after about 10s it either stop or turn.
Once the car has stopped and you press start again it will react as expected.

Have you debugged this? What does the front sensor read? Try to print out its value. Also, try using the `getMedianDistance` there as well.
Finally, 15 centimeters may be too close. In other words, it may be too "late" for the car to stop and you see this manifested as hitting the obstacle. Then once it's hit it, it reads bad values because it's too close to the obstacle itself. This is just a wild guess though, you'll have to debug this and see what the sensor is reading.

> Looking at the picture, once the car has gone into the first state it will continue
through state, 1, 2, 3, 4, 5, 6 and 7. It then jumps to state 10 and stops there. It
seems like it skips rows in the sketch. (this happened almost every time but not
always )

Instead of mentioning states on a drawing you should (also) refer to lines of code, ideally linking to them. What is for sure is that the microcontroller will not magically "jump" any lines.
Furthermore, you **keep repeating the same mistake** about making multiple measurement calls to sensors. I am using bold because I have mentioned this [before](https://github.com/DIT112-V20/group-07/pull/78#discussion_r422089449) (#78).

These kinds of loops are wrong:
```cpp
    while(left.readRangeContinuousMillimeters() <= LEFT_DIST && left.readRangeContinuousMillimeters() > 0){       //while there is still an obstical on the left side continue forward
         forward(TURN_SPEED);
    }

    while(right.getMedianDistance() <= RIGHT_DIST && right.getMedianDistance() > 0){                              //while there is still an obstical on the right side continue forward
         forward(TURN_SPEED);
    }
```

Every time you call a `getDistance`, `getMedianDistance` or `readRangeContinuousMillimeters` method, a **new** measurement is done which may be very well different than the previous one **and** wastes precious time, making your car respond slower to the changes in its environment (i.e. the obstacle). In the case of `getMedianDistance`, each call means 5 measurements (by default), making it a particularly "expensive" function.
Additionally, in the case of the micro-LIDAR sensor you don't have to check if it's larger than `0` since to my knowledge it does not follow the Smartcar convention of returning `0` to indicate an error.

> The car does not always turn 90 degrees, sometimes a bit more and sometimes a
little less which make the car drive into the obstacle again.

You are facing a very common problem in robotic systems. When you are making a movement there is no guarantee about where you will end up. In your case I can see two problems. Both can and should be solved in software if that is important for your business logic.
1. Overshooting can be "corrected" by implementing some kind of control algorithm that will slow down the car when it's reaching its target. Look at PID controllers.
2. The effect of bad gyroscope readings (manifested by rotating more or less than what you intend to) can be mediated by complementing the readings by other sources of data. In your case you have the odometers.

Both of these methods will increase the complexity of your software. I would suggest that you change your use-cases  and your entire algorithm in a way such as it does **not** depend on the car moving accurately.

> Okay, but when is an appropriate time then?

Almost any time we are not occupied with something else that is more important. :smile: For example... now!

> Because we would really like to have the discussion face to face and not via text becuause it is taking a bit too long to get answers and it is easy to missunderstand each other

We can arrange a Zoom meeting, but relying on synchronous discussions (i.e. those that both parties are present at the same time and have each others full attention) are very "expensive". Let's first try to solve things via the dedicated channels we have (i.e. GitHub and code reviews like this one).

Also keep in mind that in regards to code, live conversations are not necessarily more effective since we humans don't always explain what code does better than the code itself. :stuck_out_tongue:

**Comment by majawennroth** [comment]
Well it was you that told me to use the getMedianDistance insted of getDistance... 
and if the while loop is wrong, could you maybe explain how i should do it instead?!
i mean the mesurement need to be done at several times since the car is moving around the obstical and therfore not always at the same distance frome the object... 

Well since you always answer our questions with other questions in return we still feel like i would like to take this face to face. You have more experience in this than we do and if we as your students would like to talk to you because we need help, shouldn't you support that then and help us in the best way possible?? After all we're here to learn!
 
As i have also told you I am dyslexic and as of now i have spent more time understaning what you are writing and trying to get you to understand our problem in writing than actually trying to fix the issue. And if youre not happy about having a face to face discussion where we can actually talk and we might get answers to our questions i am happy to talk to both Tina and RBS about that.

@platisd

**Comment by platisd** [comment]
> Well it was you that told me to use the getMedianDistance insted of getDistance...
and if the while loop is wrong, could you maybe explain how i should do it instead?!

> Every time you call a `getDistance`, `getMedianDistance` or `readRangeContinuousMillimeters` method, a new measurement is done which may be very well different than the previous one and wastes precious time, making your car respond slower to the changes in its environment (i.e. the obstacle).

Which means that there isn't a problem with using the particular method(s). The problem is with *how* you use them. One of the correct ways would be this:

```cpp
    while(left.readRangeContinuousMillimeters() <= LEFT_DIST){
         forward(TURN_SPEED);
    }

    auto rightDistance = right.getMedianDistance();
    while(rightDistance <= RIGHT_DIST && rightDistance > 0){
         forward(TURN_SPEED);
         rightDistance = right.getMedianDistance();
    }
```
You should be using *the same variable* for your checks, not read (a new) distance multiple times.

Now, onto the bigger questions:

> Well since you always answer our questions with other questions in return we still feel like i would like to take this face to face. You have more experience in this than we do and if we as your students would like to talk to you because we need help, shouldn't you support that then and help us in the best way possible?? After all we're here to learn!

Unfortunately I will also be asking questions as a response to yours during a live conversation. In fact, this is what I will probably be mostly doing.
This is simply because I may be a bit more experienced with Smartcars but I am unfortunately not all-knowing. This means I cannot guess why things don't work as you'd expect them if I don't have enough data. And the only way I can get the data I need to provide some useful advice is... by asking all these questions. :)
Ιt's not like I have the answers and I am not giving them to you.

Another practical issue with live conversations is that if, for example, I ask you to check what a specific sensor is returning then I will have to wait for you to change the code appropriately and set up the system to make the measurement.
It will be a very inefficient use of time, you may be stressing to do things fast and to make things worse, I am afraid I won't be very useful to you either. Specifically, I trust that if you make the said measurement this will allow you to figure out *yourself* what is going wrong and you won't even need my help.

To conclude, please try to make the changes that I outlined before:
#### Problem 1 (Car bumps into the obstacle)

1. Increase `STOP_DIST`
2. See if this improves problem `1`
3. If it doesn't  try `getMedianDistance` in [L90](https://github.com/DIT112-V20/group-07/blob/a35578d3fed42d5a7509a89d3cb3a465a0b02eba/car/car.ino#L91).
4. See if this improves problem `1`.
5. If it doesn't, try printing out the sensor value to understand what is happening.

#### Problem 2 (Obstacle not detected while going around it)

1. Don't do multiple readings in [L149](https://github.com/DIT112-V20/group-07/blob/a35578d3fed42d5a7509a89d3cb3a465a0b02eba/car/car.ino#L149), [L159](https://github.com/DIT112-V20/group-07/blob/a35578d3fed42d5a7509a89d3cb3a465a0b02eba/car/car.ino#L159), [L184](https://github.com/DIT112-V20/group-07/blob/a35578d3fed42d5a7509a89d3cb3a465a0b02eba/car/car.ino#L184).
4. See if these improve problem `2`
5. If it doesn't insert multiple print outs in the different states of your code to determine *what* is being read and *when* 

Once you go through these steps and the problems persist, please post the findings of your investigation and we will see how we take it from there.

File: car/car.ino [hunk 1]
```diff
@@ -129,3 +122,71 @@ void handleInput() { //handle serial input (String!!)
     }  
   }
 }
+
+void driveAroundObstical(){
+  int leftDistance = left.readRangeContinuousMillimeters();
+  int rightDistance = right.getMedianDistance();
+  
+  if (leftDistance > LEFT_DIST && rightDistance > RIGHT_DIST){               //if there is no obstical on either sides rotate 90 degrees to the right
```
Comment (platisd) [inline car/car.ino]: L130 does not do what the comment says it does. If  there are no obstacles you may very well get `0` as the `rightDistance`

File: car/car.ino [hunk 2]
```diff
@@ -129,3 +122,71 @@ void handleInput() { //handle serial input (String!!)
...
+    rotateOnSpot(TURN_AROUND, TURN_SPEED);                                  //it rotates 180 degrees to the right.
+  } 
+}
+
+void obsticalTurnRight(){  //this method makes the car go around an obstical on the right side
+    rotateOnSpot(TURN_RIGHT, TURN_SPEED);
+    leftOdometer.reset();
+    rightOdometer.reset();
+
+    while(left.readRangeContinuousMillimeters() <= LEFT_DIST && left.readRangeContinuousMillimeters() > 0){      //while there is still an obstical on the left side continue forward
```
Comment (platisd) [inline car/car.ino]: Replace with:

```cpp
while(left.readRangeContinuousMillimeters() <= LEFT_DIST)
```
Same below.

File: car/car.ino [hunk 3]
```diff
@@ -129,3 +122,71 @@ void handleInput() { //handle serial input (String!!)
...
+         forward(TURN_SPEED);
+    }
+    goDistance(15, TURN_SPEED);
+    stop();
+    const long odometerLength = ((leftOdometer.getDistance() + rightOdometer.getDistance())/2);                 //this distance it the distance the car moved sideways
```
Comment (platisd) [inline car/car.ino]: You can replace this by `car.getDistance()`

File: car/car.ino [hunk 4]
```diff
@@ -129,3 +122,71 @@ void handleInput() { //handle serial input (String!!)
...
+    goDistance(odometerLength, TURN_SPEED);             //this distance it the distance the car moved sideways (used to go back in the same path but on the other side of the obstical.)
+    rotateOnSpot(TURN_RIGHT, TURN_SPEED);
+}
+
+void obsticalTurnLeft(){    //this method makes the car go around an obstical on the left side
+    rotateOnSpot(TURN_LEFT, TURN_SPEED); 
+    leftOdometer.reset();
+    rightOdometer.reset();
+
+    while(right.getMedianDistance() <= RIGHT_DIST && right.getMedianDistance() > 0){                              //while there is still an obstical on the right side continue forward
```
Comment (platisd) [inline car/car.ino]: Replace the calls to `getMedianDistance` as instructed in the pull request. Same below.

File: car/car.ino [hunk 5]
```diff
@@ -129,3 +122,71 @@ void handleInput() { //handle serial input (String!!)
...
+         forward(TURN_SPEED);
+    }
+    goDistance(15, TURN_SPEED);
+    stop();
+    const long odometerLength = ((leftOdometer.getDistance() + rightOdometer.getDistance())/2);                   //this distance it the distance the car moved sideways
```
Comment (platisd) [inline car/car.ino]: `car.getDistance()`

**Pull request closed without merging**